sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSource.java

```diff
@@ -75,167 +91,174 @@
+
+  public static <T, K, V, W extends FileInputFormat<K, V>> HDFSFileSource<T, K, V>
+  from(String filepattern,
+       Class<W> formatClass,
+       Coder<T> coder,
+       SerializableFunction<KV<K, V>, T> inputConverter) {
+    return new HDFSFileSource<>(filepattern, formatClass, coder, inputConverter, null, null, true);
   }
 
-  /**
-   * Create a {@code HDFSFileSource} based on a file or a file pattern specification.
-   */
-  protected HDFSFileSource(String filepattern,
-                           Class<? extends FileInputFormat<?, ?>> formatClass, Class<K> keyClass,
-                           Class<V> valueClass) {
-    this(filepattern, formatClass, keyClass, valueClass, null);
+  public static <K, V, W extends FileInputFormat<K, V>> HDFSFileSource<KV<K, V>, K, V>
+  from(String filepattern,
+       Class<W> formatClass,
+       Class<K> keyClass,
+       Class<V> valueClass) {
+    KvCoder<K, V> coder = KvCoder.of(getDefaultCoder(keyClass), getDefaultCoder(valueClass));
+    SerializableFunction<KV<K, V>, KV<K, V>> inputConverter =
+        new SerializableFunction<KV<K, V>, KV<K, V>>() {
+          @Override
+          public KV<K, V> apply(KV<K, V> input) {
+            return input;
+          }
+        };
+    return new HDFSFileSource<>(filepattern, formatClass, coder, inputConverter, null, null, true);
   }
 
-  /**
-   * Create a {@code HDFSFileSource} based on a single Hadoop input split, which won't be
-   * split up further.
-   */
-  protected HDFSFileSource(String filepattern,
-                           Class<? extends FileInputFormat<?, ?>> formatClass, Class<K> keyClass,
-                           Class<V> valueClass, SerializableSplit serializableSplit) {
-    this.filepattern = filepattern;
-    this.formatClass = formatClass;
-    this.keyClass = keyClass;
-    this.valueClass = valueClass;
-    this.serializableSplit = serializableSplit;
+  public static HDFSFileSource<String, LongWritable, Text>
+  fromText(String filepattern) {
+    SerializableFunction<KV<LongWritable, Text>, String> inputConverter =
+        new SerializableFunction<KV<LongWritable, Text>, String>() {
+          @Override
+          public String apply(KV<LongWritable, Text> input) {
+            return input.getValue().toString();
+          }
+        };
+    return from(filepattern, TextInputFormat.class, StringUtf8Coder.of(), inputConverter);
   }
 
-  public String getFilepattern() {
-    return filepattern;
+  public static <T> HDFSFileSource<T, AvroKey<T>, NullWritable>
+  fromAvro(String filepattern, final AvroCoder<T> coder) {
+    Class<AvroKeyInputFormat<T>> formatClass = castClass(AvroKeyInputFormat.class);
+    SerializableFunction<KV<AvroKey<T>, NullWritable>, T> inputConverter =
+        new SerializableFunction<KV<AvroKey<T>, NullWritable>, T>() {
+          @Override
+          public T apply(KV<AvroKey<T>, NullWritable> input) {
+            try {
+              return CoderUtils.clone(coder, input.getKey().datum());
```

Comment:
Yup, missed the `instanceof Writable` there.
It is up to anyone implementing a "native" `to/from` to know of Object reuse, which is fine, though I wish there was a way to clone on-read, but the coding information is obfuscated as the input is `T` at that point...